File: python-pytest/tests/varar/library.steps.py
from datetime import datetime

from library_example import FEE_PER_DAY, add_money, gbp, late_fee, may_borrow
from varar import steps


def to_date(raw):
    return datetime.strptime(raw, "%B %d, %Y").date()


def format_date(d):
    return f"{d:%B} {d.day}, {d.year}"


def to_money(raw):
    return gbp(float(raw[:-1]) / 100) if raw.endswith("p") else gbp(float(raw[1:]))


def format_money(m):
    return f"{round(m.value * 100)}p" if m.value < 1 else f"£{m.value:.2f}"


param, stimulus, sensor = steps(lambda: {"loans": (), "fee": gbp(0), "granted": False})
param("date", r"[A-Z][a-z]+ \d{1,2}, \d{4}", parse=to_date, format=format_date)
param(
    "money",
    r"£(?=.*\d.*)[-+]?\d*(?:\.(?=\d.*))?\d*|\d+p",
    parse=to_money,
    format=format_money,
)


@stimulus("borrowed {emph}, due back on {date}")
def _(state, title, due):
    return {**state, "loans": (*state["loans"], {"title": title, "due": due})}


@stimulus("returns it on {date}")
def _(state, returned_on):
    fee = gbp(0)
    for loan in state["loans"]:
        fee = add_money(fee, late_fee(loan, returned_on))
    return {**state, "fee": fee}


@sensor("owes a {money} late fee")
def _(state, expected):
    return state["fee"]


@sensor("{money} for each day overdue")
def _(state, expected):
    return FEE_PER_DAY


@stimulus("asks to borrow {emph} on {date}")
def _(state, title, on):
    return {**state, "granted": may_borrow(state["loans"], on)}


@sensor("the library refuses")
def _(state):
    if state["granted"]:
        raise AssertionError("expected the library to refuse")


@sensor("the library agrees")
def _(state):
    if not state["granted"]:
        raise AssertionError("expected the library to agree")

File: ruby-minitest/test/varar/library.steps.rb
# frozen_string_literal: true

require 'date'
require 'varar'
require_relative '../../lib/library'

to_date = ->(raw) { Date.strptime(raw, '%B %d, %Y') }
format_date = ->(d) { "#{d.strftime('%B')} #{d.day}, #{d.year}" }

to_money = ->(raw) { raw.end_with?('p') ? Library.gbp(raw[0...-1].to_f / 100) : Library.gbp(raw[1..].to_f) }
format_money = ->(m) { m.value < 1 ? "#{(m.value * 100).round}p" : format('£%.2f', m.value) }

steps(-> { { loans: [], fee: Library.gbp(0), granted: false } }) do
  param('date', '[A-Z][a-z]+ \d{1,2}, \d{4}', parse: to_date, format: format_date)
  param('money', '£(?=.*\d.*)[-+]?\d*(?:\.(?=\d.*))?\d*|\d+p', parse: to_money, format: format_money)

  stimulus('borrowed {emph}, due back on {date}') do |state, title, due|
    state.merge(loans: state[:loans] + [{ title: title, due: due }])
  end

  stimulus('returns it on {date}') do |state, returned_on|
    fee = state[:loans].reduce(Library.gbp(0)) do |acc, loan|
      Library.add_money(acc, Library.late_fee(loan, returned_on))
    end
    state.merge(fee: fee)
  end

  sensor('owes a {money} late fee') { |state, _expected| state[:fee] }

  sensor('{money} for each day overdue') { |_state, _expected| Library::FEE_PER_DAY }

  stimulus('asks to borrow {emph} on {date}') do |state, _title, on|
    state.merge(granted: Library.may_borrow(state[:loans], on))
  end

  sensor('the library refuses') { |state| raise 'expected the library to refuse' if state[:granted] }

  sensor('the library agrees') { |state| raise 'expected the library to agree' unless state[:granted] }
end
